cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -213,31 +217,83 @@ protected void shutDown() throws Exception {
    * @return list of all applications in the namespace that satisfy the specified predicate
    */
   public List<ApplicationDetail> getApps(NamespaceId namespace,
-                                         Predicate<ApplicationDetail> predicate) throws Exception {
+      Predicate<ApplicationDetail> predicate) throws Exception {
 
-    Map<ApplicationId, ApplicationSpecification> appSpecs = new LinkedHashMap<>();
-    for (ApplicationSpecification appSpec : store.getAllApplications(namespace)) {
-      appSpecs.put(namespace.app(appSpec.getName(), appSpec.getAppVersion()), appSpec);
+    List<ApplicationDetail> result = new ArrayList<>();
+    scanApplications(namespace, predicate, d -> result.add(d));
+    return result;
+  }
+
+  /**
+   * Scans all applications in the specified namespace, filtered to only include applications with an artifact name
+   * in the set of specified names and an artifact version equal to the specified version. If the specified set
+   * is empty, no filtering is performed on artifact name. If the specified version is null, no filtering is done
+   * on artifact version.
+   *
+   * @param namespace the namespace to scan apps from
+   * @param artifactNames the set of valid artifact names. If empty, all artifact names are valid
+   * @param artifactVersion the artifact version to match. If null, all artifact versions are valid
+   * @param consumer a {@link Consumer} to consume each ApplicationDetail being scanned
+   */
+  public void scanApplications(NamespaceId namespace, Set<String> artifactNames,
+      @Nullable String artifactVersion,
+      Consumer<ApplicationDetail> consumer) {
+
+    Predicate<ApplicationDetail> predicate = getAppPredicate(artifactNames, artifactVersion);
+    scanApplications(namespace, predicate, consumer);
+  }
+
+  public void scanApplications(NamespaceId namespace, Predicate<ApplicationDetail> predicate,
+      Consumer<ApplicationDetail> consumer) {
+    accessEnforcer.enforceOnParent(EntityType.DATASET, namespace,
+        authenticationContext.getPrincipal(), StandardPermission.LIST);
+
+    try (
+        BatchingConsumer<Entry<ApplicationId, ApplicationSpecification>> batchingConsumer = new BatchingConsumer<>(
+            list -> processApplications(list, predicate, consumer), 20
```

Comment:
Introduce a constant for "20". Better configurable through CConfiguration, but at least a class constant